crates/amaru/src/stages/pull_p2p.rs

```diff
@@ -0,0 +1,348 @@
+                        peer: pallas_to_amaru_peer(&pid),
+                        raw_header: header.cbor,
+                        point: todo!(),
+                        span: Span::current(),
+                    }),
+                    network_cmd: Some(InitiatorCommand::ContinueSync(pid.clone())),
+                    ..Default::default()
+                };
+
+                WorkSchedule::Unit(work)
+            }
+
+            // This is triggered when a rollback has been received from a peer that is performing
+            // chain-sync. Comments from the block header received event apply here as well.
+            InitiatorEvent::RollbackReceived(pid, point, _) => {
+                tracing::info!(%pid, "rollback received");
+
+                let work = WorkUnit {
+                    sync_data: Some(ChainSyncEvent::Rollback {
+                        peer: pallas_to_amaru_peer(&pid),
+                        rollback_point: pallas_to_amaru_point(&point),
+                        span: Span::current(),
+                    }),
+                    network_cmd: Some(InitiatorCommand::ContinueSync(pid.clone())),
+                    ..Default::default()
+                };
+
+                WorkSchedule::Unit(work)
+            }
+
+            // This is triggered when a block body has been received from a peer that is handling a
+            // block-fetch request. The network machinery explicitly avoids decoding the body for
+            // performance reasons. It's up for us (amaru) to decide when / how to execute the
+            // decoding.
+            //
+            // The downside of this approach is that we don't have much information about
+            // the block at this point, just the payload. We can't infer which point in the chain or
+            // to which downstream request it belongs to. The behavior here is to just dispatch the
+            // block body downstream and let downstream stages decide how to handle it.
+            //
+            // There's no need for explicit commands back to the network machinery since the
+            // continuation of the mini-protocol is unambiguous. If the fetch request involved more
+            // blocks, the network machinery will keep receiving and notifying us of new
+            // block bodies.
+            InitiatorEvent::BlockBodyReceived(pid, body) => {
+                tracing::info!(%pid, "block body received");
+
+                let work = WorkUnit {
+                    block_data: Some(body),
+                    ..Default::default()
+                };
+
+                WorkSchedule::Unit(work)
+            }
+
+            // There're other events that the network can trigger, but we don't care about them for
+            // now.
+            _ => WorkSchedule::Idle,
+        }
+    }
+
+    fn handle_request(&self, request: &DownstreamRequest) -> WorkSchedule<WorkUnit> {
+        match request {
+            DownstreamRequest::FetchBlock(range) => {
```

Comment:
How do we know which peer to send the request to? This seems important because there are some security implications in a peer not being able to deliver a block it's supposed to have (eg. they sent a header but fail to deliver block)